smithyctl/internal/command/workflow/parse.go

```diff
@@ -0,0 +1,368 @@
+		return nil, errors.Errorf("failed to resolve absolute component path: %w", err)
+	}
+
+	c, err := sp.componentParser.Parse(ap)
+	if err != nil {
+		return nil, errors.Errorf("failed to parse local component: %v", err)
+	}
+	return c, nil
+}
+
+func parseRawWorkflowSpec(path string) (*workflowAdapter, error) {
+	const (
+		defaultSmithyWorkflowFileNameYaml = "workflow.yaml"
+		defaultSmithyWorkflowFileNameYml  = "workflow.yml"
+	)
+
+	if !strings.HasSuffix(path, defaultSmithyWorkflowFileNameYaml) && !strings.HasSuffix(path, defaultSmithyWorkflowFileNameYml) {
+		return nil, errors.Errorf(
+			"invalid file path %s, has to either point to a component file",
+			path,
+		)
+	}
+
+	b, err := os.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errors.Errorf("%s does not exist", path)
+		}
+		return nil, fmt.Errorf("failed to read workflow config file: %w", err)
+	}
+
+	var workflow workflowAdapter
+	if err := yaml.NewDecoder(bytes.NewReader(b)).Decode(&workflow); err != nil {
+		return nil, fmt.Errorf("failed to decode file '%s': %w", path, err)
+	}
+
+	return &workflow, workflow.validate()
+}
+
+// parseOverrides reads the overrides from the passed path and returns them organised
+// by component name and parameter name.
+func parseOverrides(path string) (map[string]map[string]v1.Parameter, error) {
+	var overrides = make(map[string]map[string]v1.Parameter)
+	// If the overrides haven't been specified, return early.
+	if path == "" {
+		return overrides, nil
+	}
+
+	b, err := os.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errors.Errorf("%s does not exist", path)
+		}
+		return nil, fmt.Errorf("failed to read overrides file: %w", err)
+	}
+
+	var tmpOverrides map[string][]v1.Parameter
+	if err := yaml.NewDecoder(bytes.NewReader(b)).Decode(&tmpOverrides); err != nil {
+		return nil, fmt.Errorf("failed to decode file overrides '%s': %w", path, err)
+	}
+
+	for componentName, params := range tmpOverrides {
+		overrides[componentName] = make(map[string]v1.Parameter)
+		for _, param := range params {
+			if err := param.Validate(); err != nil {
+				return nil, errors.Errorf("failed to validate override parameter for component '%s': %w", componentName, err)
+			}
+			overrides[componentName][param.Name] = param
+		}
+	}
+
+	return overrides, nil
+}
+
+// filterComponents returns the components that match the passed component type.
+func filterComponents(components []*v1.Component, componentType v1.ComponentType) []v1.ComponentRef {
+	var filteredComponents []v1.ComponentRef
+	for _, c := range components {
+		if c.Type == componentType {
+			filteredComponents = append(
+				filteredComponents,
+				v1.ComponentRef{
+					Component: *c,
+				},
+			)
+		}
+	}
+	return filteredComponents
+}
+
+// renderComponents templates parameters in component.yaml files.
+// The parameters are always specified with a '.parameters' prefix.
+func renderComponents(components []*v1.Component) error {
```

Comment:
Copy pasta from SAAS impl minus adding more custom functions for shlex etc - will come later